### ISSUE TITLE
導入K-means 進行資料預處理

### ISSUE THREAD

**Issue by JNRLEE** [JNRLEE/MicForDysphasiaOnCloud#2]
目標

實現資料的前置處理，利用 K-means 聚類將不同長度的資料轉換為固定數量的 cluster，統一輸入結構，適配後續模型（例如 CNN 或 Transformer）。

背景
	1.	資料特性：
	•	資料為三維陣列，形狀為 [1, 512, x]，其中 x（時間步長）在不同樣本中不一致（範圍約為 1000-10000）。
	•	不同樣本的長度變化使得直接輸入深度學習模型（如 CNN）困難。
	2.	解決方法：
	•	將每個樣本的時間步特徵通過 K-means 聚類轉換為固定數量的 cluster。
	•	聚類結果可以表示為直方圖（頻率分佈），統一輸出形狀。

實現步驟
	1.	匯入數據：
	•	加載樣本數據，形狀為 [1, 512, x]。
	•	處理不同時間步長的樣本。
	2.	標準化數據：
	•	對每個樣本的特徵數據（512 維）進行標準化，以適應 K-means 聚類。
	3.	應用 K-means 聚類：
	•	設置固定的 cluster 數（例如，n_clusters=100）。
	•	對每個樣本的時間步進行聚類，將特徵映射到 cluster 標籤。
	4.	生成固定長度表示：
	•	使用 cluster 標籤生成直方圖，將每個樣本轉換為長度為 n_clusters 的固定向量。
	5.	保存處理結果：
	•	將處理後的數據保存為 numpy 陣列或其他可用於後續分析的格式。

程式碼參考

'''
import numpy as np
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

# 模擬不同長度的資料
data_samples = [np.random.rand(1, 512, np.random.randint(1000, 10000)) for _ in range(10)]

# 設置 k-means 聚類的 cluster 數
n_clusters = 100
kmeans = KMeans(n_clusters=n_clusters, random_state=42)

processed_samples = []
for sample in data_samples:
    _, feature_dim, time_steps = sample.shape
    reshaped_sample = sample[0].T  # 轉置為 (time_steps, feature_dim)
    scaler = StandardScaler()
    normalized_sample = scaler.fit_transform(reshaped_sample)
    kmeans.fit(normalized_sample)
    cluster_labels = kmeans.labels_
    histogram, _ = np.histogram(cluster_labels, bins=n_clusters, range=(0, n_clusters))
    processed_samples.append(histogram)

processed_samples = np.array(processed_samples)  # shape: (num_samples, n_clusters)
print(f"Processed samples shape: {processed_samples.shape}")
'''

預期結果
	1.	輸出統一結構的數據：
	•	每個樣本被轉換為形狀為 (n_clusters,) 的固定長度表示。
	•	例如，如果 n_clusters=100，則輸出為 (num_samples, 100)。
	2.	改進模型適配性：
	•	統一的數據結構便於後續輸入 CNN 或其他模型進行學習。

測試用例
	1.	測試能否成功將不同長度的樣本轉換為統一形狀。
	2.	測試在輸入到模型後，數據格式是否正確。

	5.	提交 Issue：
填寫完內容後，點擊 “Submit new issue” 完成創建。

結果

Issue 創建完成後，您可以在 repository 的 “Issues” 頁面中看到這個 Issue，並根據需要分配相關人員或標籤（例如，enhancement、preprocessing）。這將幫助您更好地跟蹤和管理這個任務的進展。
